### PR TITLE
Make allowed_classes configurable when unserializing formState

### LINKED ISSUE #126: DatePicker field in multi-page forms can't get unserialized
Using a DatePicker field in a multi-page form throws an exception, when you try to re-render the field with a value set.

E.g.

1. you set a Date on the first page
2. go further to the next page
3. step back to previous page

**Exception**
``` 
Could not convert target type “DateTime”: No converter found which can be used to convert from “__PHP_Incomplete_Class” or “object” to “DateTime”.
```

The unserialize of the form state causes this, because it is restricted to `FormState` class, so the `DateTime` object is ignored and handled as incomplete class.

This behavior was introduced by this changes:
https://github.com/neos/form/commit/c186327d9ada84e48fcee05b4828e4da1a8d1e60
https://github.com/neos/form/commit/ee593c2111515f6cc6ad6a77c13e234c1be5b8f2

Affected: Neos.Form >= 5.0

See also: https://discuss.neos.io/t/datetime-conversion-in-form/5007

### LINKED ISSUE #143: Multi-Step forms fail when saving objects in the Form State
In my usecase the FormState holds asset objects like `Neos\Media\Domain\Model\Document` or `Neos\Media\Domain\Model\Image`. When form is rendered with initalized formState having one of the mentioned objects, an exception is thrown:

`
Absorbed Exception: Notice: property_exists(): The script tried to execute a method or access a property of an incomplete object. Please ensure that the class definition &quot;Neos\Media\Domain\Model\Document&quot; of the object you are trying to operate on was loaded _before_ unserialize() gets called or provide an autoloader to load the class definition in .../Data/Temporary/Development/Cache/Code/Flow_Object_Classes/Neos_Flow_Persistence_Doctrine_PersistenceManager.php line 188
`

Is related to #126

### PR DESCRIPTION
**Pull request by ger4003**
With a new configuration option it is possible to add class names to `allowed_classes ` for `unserialize()` formState within `FormRuntime::initializeFormStateFromRequest`

closes #143

### CODE REVIEW

File: Classes/Core/Runtime/FormRuntime.php [hunk 1]
```diff
@@ -181,7 +188,12 @@ protected function initializeFormStateFromRequest()
             $this->formState = new FormState();
         } else {
             $serializedFormState = $this->hashService->validateAndStripHmac($serializedFormStateWithHmac);
-            $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => [FormState::class, \DateTime::class, \DateTimeImmutable::class]]);
+            $allowedClassesInFormState = Arrays::removeEmptyElementsRecursively($this->allowedClassesInFormState);
+            if (is_array($allowedClassesInFormState) && !empty($allowedClassesInFormState)) {
+                $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => $this->allowedClassesInFormState]);
```
Comment (bwaidelich) [inline Classes/Core/Runtime/FormRuntime.php:193]: you introduce `$allowedClassesInFormState` but don't use it here :)

Comment (bwaidelich) [reply Classes/Core/Runtime/FormRuntime.php:193]:
```suggestion
                $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => $allowedClassesInFormState]);
```

File: Classes/Core/Runtime/FormRuntime.php [hunk 2]
```diff
@@ -181,7 +188,12 @@ protected function initializeFormStateFromRequest()
...
+            } else {
+                $this->formState = unserialize(base64_decode($serializedFormState));
```
Comment (bwaidelich) [inline Classes/Core/Runtime/FormRuntime.php:195]: This fallback to an unsafe mode seems dangerous to me.
Couldn't we just simplify this to a single line:
```php
$this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => array_filter($this->allowedClassesInFormState)]);
```
?

Comment (ger4003) [reply Classes/Core/Runtime/FormRuntime.php:195]: Yes we can. It is  much better solution, thanks.

**Review by bwaidelich — APPROVED** [review]
+1 by reading. Thanks you, @ger4003 !

**Comment by bwaidelich** [comment]
I'm just not sure about the target branch. IIRC features can only go into a `5.3` release!?

**Comment by ger4003** [comment]
I can change the target branch to `master`, that is fine by me. @bwaidelich  Should I?

**Comment by bwaidelich** [comment]
> I can change the target branch to master

yes please. Thanks

**Comment by bwaidelich** [comment]
@ger4003 Just a little update: I'm currently reviewing this and the change works great, but we are currently discussing whether the `allowed_classes` option could be skipped altogether. I will keep you updated

**Review by bwaidelich — CHANGES_REQUESTED** [review]
And here is the update:
I think this one is obsolete with #145 and can be closed.
Thanks again and sorry for the inconvenience!

**Comment by ger4003** [comment]
#close in favor of #145

**Pull request closed without merging**